lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,323 @@
+        Automatically assumes this.
+        """
+        if self._prepared:
+            raise Exception("Encoder is already prepared.", flush=True)
+
+        # TODO: Make tokenizer custom with partial function; feed custom->model
+        if self._tokenizer is None:
+            self._tokenizer = self._tokenizer_class.from_pretrained(
+                self._pretrained_model_name
+            )
+
+        # Replaces empty strings with ''
+        priming_data = [x if x is not None else "" for x in priming_data]
+
+        # Checks training data details
+        # TODO: The dict should be streamlined
+        output_avail = training_data is not None and len(training_data["targets"]) == 1
+
+        output_type = training_data["targets"][0]["output_type"]
+
+        type_flag = output_type == COLUMN_DATA_TYPES.CATEGORICAL
+
+        if self._custom_train and output_avail and type_flag:
+            print("Training model.", flush=True)
+
+            # Prepare priming data into tokenized form + attention masks
+            text = self._tokenizer(priming_data, truncation=True, padding=True)
+
+            print("\tOutput trained is categorical", flush=True)
+
+            if training_data["targets"][0]["encoded_output"].shape[1] > 1:
+                labels = training_data["targets"][0]["encoded_output"].argmax(
+                    dim=1
+                )  # Nbatch x N_classes
+            else:
+                labels = training_data["targets"][0]["encoded_output"]
+
+            label_size = len(set(training_data["targets"][0]["unencoded_output"])) + 1
+
+            # Construct the model
+            self._model = self._classifier_model_class.from_pretrained(
+                self._pretrained_model_name,
+                num_labels=label_size,
+            ).to(self.device)
+
+            # Construct the dataset for training
+            xinp = TextEmbed(text, labels)
+            dataset = DataLoader(xinp, batch_size=self._batch_size, shuffle=True)
+
+            # If max length not set, adjust
+            if self._max_len is None:
+                self._max_len = self._model.config.max_position_embeddings
+
+            if self._frozen:
+                print("\tFrozen Model + Training Classifier Layers", flush=True)
+                """
+                Freeze the base transformer model and train
+                a linear layer on top
+                """
+                # Freeze all the transformer parameters
+                for param in self._model.base_model.parameters():
+                    param.requires_grad = False
+
+                optimizer_grouped_parameters = self._model.parameters()
+
+            else:
+                print("\tFine-tuning model", flush=True)
+                """
+                Fine-tuning parameters with weight decay
+                """
+                no_decay = [
+                    "bias",
+                    "LayerNorm.weight",
+                ]  # decay on all terms EXCLUDING bias/layernorms
```

Comment:
Maybe a small comment explaining why? (Unless it's just heuristic for benchmarking)

Reply:
This is sort of the 'prescribed recipe' for fine-tuning; the goal though is to dampen the weight updates so they don't overfit dramatically to the new dataset while tuning. I'm not particularly sure why biases/layer-norms are omitted, though